### PR TITLE
feat(algorithm): CLI implementation

### PR DESCRIPTION
**Pull request by amenes12**
### What does it do

Implements the base algorithm to calculate cash flows and profitability indicators of a leasing operation.

### Why is it needed

It's the basic implementation of the main algorithm which will do the calculations in our app. We can easily link the UI with this logic.

### CODE REVIEW

File: algorithm/index.mjs [hunk 1]
```diff
@@ -1,17 +1,487 @@
...
+
+  console.log(`\nN° cuotas al año: ${cuotasAnuales}`);
+  console.log(`N° periodos de pago: ${periodos}`);
+  console.log(`Frecuencia de pago: ${inputs.paymentFrequency} días\n`);
+
+  console.log(`Tipo de tasa de interés: ${inputs.interestRateType}`);
+  console.log(`Tasa efectiva del periodo: ${periodicalInterestRate * 100}%\n`);
+
+  console.log(`Tasa descuento Ks equivalente: ${descuentoKs * 100}%`);
+  console.log(`Tasa descuento WACC equivalente: ${descuentoWACC * 100}%`);
```
Comment (dalbitresb12) [inline algorithm/index.mjs]: Could you make sure that you use the `Intl.NumberFormat` API also for printing all the results?

File: algorithm/index.mjs [hunk 2]
```diff
@@ -1,17 +1,487 @@
...
+    message: "Tasa de descuento Ks: ",
+    validate: percentageValidation,
+    filter: value => {
+      const parsed = Number(value);
+      if (percentageValidation(parsed) === true) {
+        return parsed;
+      }
+      return value;
+    },
+    transformer: input => `${input}%`,
```
Comment (dalbitresb12) [inline algorithm/index.mjs]: Use the percentage formatter everywhere you can.

File: algorithm/index.mjs [hunk 3]
```diff
@@ -1,17 +1,487 @@
...
+    message: "Tasa de descuento WACC: ",
+    validate: percentageValidation,
+    filter: value => {
+      const parsed = Number(value);
+      if (percentageValidation(parsed) === true) {
+        return parsed;
+      }
+      return value;
+    },
+    transformer: input => `${input}%`,
```
Comment (dalbitresb12) [inline algorithm/index.mjs]: Use the percentage formatter everywhere you can.

File: algorithm/index.mjs [hunk 4]
```diff
@@ -1,17 +1,487 @@
...
+      amortizacion = cuota - intereses;
+      saldoFinal = saldoInicial + amortizacion;
+    }
+
+    console.log(`\nSaldo inicial: S/ ${roundMoney(saldoInicial)}`);
+    console.log(`Intereses: S/ ${roundMoney(intereses)}`);
+    console.log(`Cuota: S/ ${roundMoney(cuota)}`);
+    console.log(`Amortización: S/ ${roundMoney(amortizacion)}`);
+    console.log(`Monto de seguro contra todo riesgo: S/ ${roundMoney(insuranceAmount)}`);
+    console.log(`Costos periódicos: S/. ${roundMoney(periodicCosts)}`);
```
Comment (dalbitresb12) [inline algorithm/index.mjs]: You can use the `Intl.NumberFormat` API here as well.

File: algorithm/index.mjs [hunk 5]
```diff
@@ -1,17 +1,487 @@
...
+
+    ahorroTributario = (intereses + insuranceAmount + periodicCosts + depreciacion) * IR;
+    montoIGVPeriodico = (cuota + insuranceAmount + periodicCosts) * IGV;
+    flujoBruto = cuota + insuranceAmount + periodicCosts;
+
+    if (inputs.buyingOption && periodo === periodos) {
+      montoIGVPeriodico = (cuota + insuranceAmount + periodicCosts + montoRecompra) * IGV;
+      flujoBruto = cuota + insuranceAmount + periodicCosts + montoRecompra;
+      console.log(`Monto de recompra: S/ ${roundMoney(montoRecompra)}`);
```
Comment (dalbitresb12) [inline algorithm/index.mjs]: You can use the `Intl.NumberFormat` API here as well.

File: algorithm/index.mjs [hunk 6]
```diff
@@ -1,17 +1,487 @@
...
+    vnaFlujoBruto += flujoBruto / Math.pow(1 + descuentoKs, periodo);
+    vnaFlujoNeto += flujoNeto / Math.pow(1 + descuentoWACC, periodo);
+
+    console.log(`Saldo final: S/ ${roundMoney(saldoFinal)}`);
+    console.log(`Depreciación: S/ ${roundMoney(depreciacion)}`);
+    console.log(`Ahorro tributario: S/ ${roundMoney(ahorroTributario)}`);
+    console.log(`IGV: S/ ${roundMoney(montoIGVPeriodico)}`);
+    console.log(`Flujo Bruto: S/ ${roundMoney(flujoBruto)}`);
+    console.log(`Flujo con IGV: S/ ${roundMoney(flujoIGV)}`);
+    console.log(`Flujo Neto:S/ ${roundMoney(flujoNeto)}\n`);
```
Comment (dalbitresb12) [inline algorithm/index.mjs]: You can use the `Intl.NumberFormat` API here as well.

File: algorithm/index.mjs [hunk 7]
```diff
@@ -1,17 +1,487 @@
...
+    flujosNetos.push(roundMoney(flujoNeto));
+  }
+
+  const desembolsoTotal = totalIntereses + totalAmortizacion + totalSeguro + totalCostosPeriodicos + montoRecompra;
+
+  console.log(`Monto total por intereses: S/ ${roundMoney(-totalIntereses)}`);
+  console.log(`Monto total amortizado: S/ ${roundMoney(-totalAmortizacion)}`);
+  console.log(`Monto total por costos periódicos: S/ ${roundMoney(-totalCostosPeriodicos)}`);
+  console.log(`Monto total por seguro contra todo riesgo: S/ ${roundMoney(-totalSeguro)}`);
+  console.log(`Desembolso total: S/ ${roundMoney(-desembolsoTotal)}\n`);
```
Comment (dalbitresb12) [inline algorithm/index.mjs]: You can use the `Intl.NumberFormat` API here as well.

File: algorithm/index.mjs [hunk 8]
```diff
@@ -1,17 +1,487 @@
...
+  const tceaFlujoBruto = Math.pow(1 + tirFlujoBruto, cuotasAnuales) - 1;
+  const tceaFlujoNeto = Math.pow(1 + tirFlujoNeto, cuotasAnuales) - 1;
+
+  console.log("= = = = = Indicadores de Rentabilidad = = = = =");
+  console.log(`TCEA Flujo Bruto: ${tceaFlujoBruto * 100}%`);
+  console.log(`TCEA Flujo Neto: ${tceaFlujoNeto * 100}%`);
+  console.log(`VAN Flujo Bruto: ${roundMoney(vnaFlujoBruto)}`);
+  console.log(`VAN Flujo Neto: ${roundMoney(vnaFlujoNeto)}`);
+  console.log(`TIR Flujo Bruto: ${tirFlujoBruto * 100}%`);
+  console.log(`TIR Flujo Neto: ${tirFlujoNeto * 100}%`);
```
Comment (dalbitresb12) [inline algorithm/index.mjs]: You can use the `Intl.NumberFormat` API here as well.

File: algorithm/index.mjs [hunk 9]
```diff
@@ -1,17 +1,487 @@
...
+
+    intereses = -(periodicalInterestRate * saldoInicial);
+
+    if (periodo < periodos) {
+      const input = await inquirer.prompt({
+        name: "plazoDeGracia",
+        type: "list",
+        message: `Escoja el tipo de plazo de gracia: `,
+        choices: ["Total", "Parcial", "Sin plazo de gracia"],
+      });
```
Comment (dalbitresb12) [inline algorithm/index.mjs]: Is this better than trying to implement the [table prompt](https://github.com/eduardoboucas/inquirer-table-prompt) we talked about in Telegram?

**Review by dalbitresb12 — CHANGES_REQUESTED** [review]
Couple of minor things.

Something that would be great is if all the logic behind each calculated value could be abstracted into its own functions and modules so that we can reuse those implementations (just adding type safety with TypeScript later) for the web frontend.

File: algorithm/index.mjs (hunk 8 above)
Comment (amenes12) [reply algorithm/index.mjs]: Done

File: algorithm/index.mjs (hunk 7 above)
Comment (amenes12) [reply algorithm/index.mjs]: Done

File: algorithm/index.mjs (hunk 3 above)
Comment (amenes12) [reply algorithm/index.mjs]: Done

File: algorithm/index.mjs (hunk 4 above)
Comment (amenes12) [reply algorithm/index.mjs]: Done

File: algorithm/index.mjs (hunk 5 above)
Comment (amenes12) [reply algorithm/index.mjs]: Done

File: algorithm/index.mjs (hunk 6 above)
Comment (amenes12) [reply algorithm/index.mjs]: Done

File: algorithm/index.mjs (hunk 1 above)
Comment (amenes12) [reply algorithm/index.mjs]: Done

File: algorithm/index.mjs (hunk 2 above)
Comment (amenes12) [reply algorithm/index.mjs]: Done

**Review by dalbitresb12 — APPROVED** [review]
LGTM

**Pull request merged**